### ISSUE TITLE
Steep 1.7.0 stuck when running check

### ISSUE THREAD

**Issue by elct9620** [soutaro/steep#1162]
After upgrading to 1.7.0 the `check` command will stuck when getting this error message

<details>

```
[Steep 1.7.0] [typecheck:typecheck@0] [background] Unexpected error: #<RuntimeError: self_type cannot include 'instance' type: ::FFI::_DataConverter[::FFI::Pointer, instance, untyped]>
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/interface/builder.rb:40:in `validate_fvs'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/interface/builder.rb:27:in `validate'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/interface/builder.rb:13:in `initialize'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:413:in `new'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:413:in `block (2 levels) in check_type0'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/relation.rb:89:in `map'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:410:in `block in check_type0'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/result.rb:57:in `initialize'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/result.rb:301:in `new'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/result.rb:301:in `Expand'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:408:in `check_type0'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:209:in `block (2 levels) in check_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:37:in `push_assumption'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:208:in `block in check_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:135:in `block in tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:39:in `tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:135:in `tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:198:in `check_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:187:in `block in check'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:26:in `with_context'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:186:in `check'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:75:in `block in validate_type_application_constraints'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:66:in `each'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:66:in `validate_type_application_constraints'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:127:in `validate_type_application'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:140:in `validate_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:229:in `block (3 levels) in validate_definition_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/rbs-3.5.1/lib/rbs/types.rb:1040:in `each_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/rbs-3.5.1/lib/rbs/method_type.rb:86:in `each_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:228:in `block (2 levels) in validate_definition_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/subtyping/check.rb:53:in `push_variable_bounds'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:227:in `block in validate_definition_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:199:in `block (2 levels) in each_method_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:198:in `each'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:198:in `block in each_method_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:196:in `each_value'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:196:in `each_method_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:222:in `validate_definition_type'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:389:in `block (3 levels) in validate_one_class_decl'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   <internal:kernel>:90:in `tap'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:301:in `block (2 levels) in validate_one_class_decl'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:135:in `block in tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:39:in `tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:135:in `tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:243:in `block in validate_one_class_decl'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:556:in `rescue_validation_errors'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:240:in `validate_one_class_decl'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/signature/validator.rb:400:in `validate_one_class'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/services/type_check_service.rb:256:in `block (6 levels) in validate_signature'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep.rb:221:in `sample'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/services/type_check_service.rb:253:in `block (5 levels) in validate_signature'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/3.3.0/set.rb:501:in `each_key'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/3.3.0/set.rb:501:in `each'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/services/type_check_service.rb:252:in `block (4 levels) in validate_signature'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep.rb:256:in `measure2'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/services/type_check_service.rb:251:in `block (3 levels) in validate_signature'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/services/type_check_service.rb:225:in `each'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/services/type_check_service.rb:225:in `block (2 levels) in validate_signature'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep.rb:192:in `measure'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/services/type_check_service.rb:221:in `block in validate_signature'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:135:in `block in tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:39:in `tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:135:in `tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/services/type_check_service.rb:220:in `validate_signature'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/server/type_check_worker.rb:187:in `handle_job'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/server/base_worker.rb:56:in `block (2 levels) in run'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:135:in `block in tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:39:in `tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/activesupport-7.1.3.4/lib/active_support/tagged_logging.rb:135:in `tagged'
[Steep 1.7.0] [typecheck:typecheck@0] [background]   /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/server/base_worker.rb:46:in `block in run'
```

</details>

Pressed `Ctrl + C` then got this

```
/Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/server/base_worker.rb:91:in `join': No live threads left. Deadlock? (fatal)
2 threads, 2 sleeps current:0x000000013d604ea0 main thread:0x000000013d604ea0
* #<Thread:0x00000001003ab320 sleep_forever>
   rb_thread_t:0x000000013d604ea0 native:0x000000020454fac0 int:0

* #<Thread:0x0000000138cf0588 /Users/elct9620/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/steep-1.7.0/lib/steep/server/base_worker.rb:42 sleep_forever>
   rb_thread_t:0x0000000100bedc10 native:0x000000016fcab000 int:0
    depended by: tb_thread_id:0x000000013d604ea0
```

* Ruby version: 3.3.0
* Platform: Apple M3 (macOS 14.4.1)

> Tried `steep check --jobs 1` still stuck

It can be reproduced in this repository:
https://github.com/elct9620/hanami-lambda

* Remove `gem "steep"` in `Gemfile`
* Add `spec.add_development_dependency "steep"` to upgrade to `1.7.0`
* Run `rbs collection update`
* Run `steep check`

**Comment by soutaro** [comment]
Confirmed. I'm working to fix the `self_type cannot include 'instance' type` error (by delaying the validation) but type checking FFI still reports some problems.

```
sig/sig-ffi/ffi/auto_pointer.rbs:15:65: [error] Type application of `::FFI::AutoPointer::Releaser::_Proc` doesn't satisfy the constraints: self <: ::FFI::Pointer
│ Diagnostic ID: RBS::UnsatisfiableTypeApplication
│
└     def initialize: (Pointer pointer, Method | ^(self) -> void | Releaser::_Proc[self] callable) -> self
                                                                   ~~~~~~~~~~~~~~~~~~~~~

sig/sig-ffi/ffi/struct.rbs:23:29: [error] Type application of `::FFI::Type::Mapped` doesn't satisfy the constraints: ::FFI::Struct::ManagedStructConverter[::FFI::ManagedStruct[::FFI::AutoPointer, untyped], ::FFI::AutoPointer] <: ::FFI::_DataConverter[::FFI::Pointer, instance, untyped]
│ Diagnostic ID: RBS::UnsatisfiableTypeApplication
│
└     def self.auto_ptr: () -> Type::Mapped[
                               ~~~~~~~~~~~~~

sig/sig-ffi/ffi/struct.rbs:5:15: [error] Type application of `::FFI::Type::Mapped` doesn't satisfy the constraints: ::FFI::StructByReference[::FFI::Struct[::FFI::AbstractMemory, untyped], ::FFI::AbstractMemory] <: ::FFI::_DataConverter[::FFI::AbstractMemory, instance, untyped]
│ Diagnostic ID: RBS::UnsatisfiableTypeApplication
│
└     type ptr = Type::Mapped[
                 ~~~~~~~~~~~~~

sig/sig-ffi/ffi/library.rbs:36:45: [error] Type `::FFI::DataConverter` is generic but used as a non generic type
│ Diagnostic ID: RBS::InvalidTypeApplication
│
└     def typedef: [T < Type] (T old, Symbol | DataConverter add, ?untyped) -> T
```

(Moved the RBS files from the gem package to `sig` dir for quick testing.)

@ParadoxV5 Do you think the errors are by some bugs of Steep? Or, it may be because of the type definition itself? (The last `DataConverter` problem is true problem in the type definition.)

**Comment by ParadoxV5** [comment]
Re: FFI problems:
* ffi/ffi#1107
  * (Likely) depends on #916
  * Relates to ffi/ffi#1108

**Comment by soutaro** [comment]
This problem is `self` type in `#initialize` method. Using `instance` type in the method type solves the problem.

```
sig/sig-ffi/ffi/auto_pointer.rbs:15:65: [error] Type application of `::FFI::AutoPointer::Releaser::_Proc` doesn't satisfy the constraints: self <: ::FFI::Pointer
│ Diagnostic ID: RBS::UnsatisfiableTypeApplication
│
└     def initialize: (Pointer pointer, Method | ^(self) -> void | Releaser::_Proc[self] callable) -> self
                                                                   ~~~~~~~~~~~~~~~~~~~~~
```

The `#initialize` type is copied to `.new` method type, which means the `self` type also appears different context, `instance` for instance methods, `class` for singleton methods.

We should prohibit using `self` type in `#initialize` method types to avoid this confusion.

**Comment by ParadoxV5** [comment]
> The `#initialize` type is copied to `.new` method type, which means the `self` type also appears different context, `instance` for instance methods, `class` for singleton methods.

We should clarify this context-dependent behavior in [the RBS docs](https://github.com/ruby/rbs/blob/master/docs/syntax.md#base-types).

> We should prohibit using `self` type in `#initialize` method types to avoid this confusion.

No. We should automatically adapt uses of `self` type in `#initialize` to remedy this.

`self` and `instance` have differences, even if only intuitively but not implemented.
```ruby
o = Object.new.extend Comparable #: Object & Comparable
o.dup.is_a? Comparable #=> false (instance ➡ Object)
o.itself.is_a? Comparable #=> true (self ➡ Object & Comparable)
```

**Comment by soutaro** [comment]
Released Steep 1.7.1, that fixes the issue.

Type checking FFI type definition still fails, and I think these are the problem of the type definition itself.

The `self` types in `initialize` method type causes issues as mentioned above.

```
sig/sig-ffi/ffi/struct.rbs:23:29: [error] Type application of `::FFI::Type::Mapped` doesn't satisfy the constraints: ::FFI::Struct::ManagedStructConverter[::FFI::ManagedStruct[::FFI::AutoPointer, untyped], ::FFI::AutoPointer] <: ::FFI::_DataConverter[::FFI::Pointer, instance, untyped]
│ Diagnostic ID: RBS::UnsatisfiableTypeApplication
│
└     def self.auto_ptr: () -> Type::Mapped[
                               ~~~~~~~~~~~~~
```

This error still reports because of `Struct[...] <: ManagedStruct[...]`, which doesn't hold. (Using `ManagedStruct[AutoPointer, untyped]` instead of `instance` solves the type error, but I'm not sure if it's something the FFI gem wants.)
